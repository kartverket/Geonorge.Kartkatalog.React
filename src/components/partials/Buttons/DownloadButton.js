import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { removeItemSelectedForDownload, addItemSelectedForDownload } from '../../../actions/DownloadItemActions'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

class DownloadButton extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isSelectedForDownload: this.selectedForDownloadIsAddedToLocalStorage(this.getDownloadButton())
        };
    }

    compareItemsToDownload(itemToCompare, itemToCompareWith) {
        return itemToCompare.Uuid === itemToCompareWith.Uuid && itemToCompare.Title === itemToCompareWith.Title
    }

    selectedForDownloadIsAddedToLocalStorage(itemToCompare) {
        if (localStorage.itemsToDownload && Array.isArray(JSON.parse(localStorage.itemsToDownload))) {
            let isAddedToLocalStorage = false;
            JSON.parse(localStorage.itemsToDownload).forEach((mapItemToCompareWith) => {
                if (this.compareItemsToDownload(itemToCompare, mapItemToCompareWith)) {
                    isAddedToLocalStorage = true;
                }
            });
            return isAddedToLocalStorage;
        } else {
            return false;
        }
    }
    getDownloadButton() {
        return {
            Uuid: this.props.searchResult.Uuid,
            Title: this.props.searchResult.Title,
            DistributionProtocol: this.props.searchResult.DistributionProtocol,
            IsOpenData: this.props.searchResult.IsOpenData,
            DistributionUrl: this.props.searchResult.DistributionUrl
        }
    }
    addToDownloadList(item) {
        this.setState({
            isSelectedForDownload: true
        });
        this.props.addItemSelectedForDownload(item);
    }
    removeFromDownloadList(item) {
        this.setState({
            isSelectedForDownload: false
        });
        this.props.removeItemSelectedForDownload(item);
    }

    isGeonorgeDownload() {
        return this.props.searchResult.DistributionProtocol === 'GEONORGE:DOWNLOAD'
    }
    showDownloadLink() {
        return this.props.searchResult.DistributionUrl &&
            (this.props.searchResult.DistributionProtocol === 'WWW:DOWNLOAD-1.0-http--download' ||
                this.props.searchResult.DistributionProtocol === 'GEONORGE:FILEDOWNLOAD')
            && this.props.searchResult.Type === 'dataset'
    }

    render() {
        let button = this.getDownloadButton();
        if (this.isGeonorgeDownload(this.props.searchResult.DistributionProtocol)) {
            let action = this.state.isSelectedForDownload
                ? () => this.removeFromDownloadList(button)
                : () => this.addToDownloadList(button);
            let icon = <FontAwesomeIcon icon={this.state.isSelectedForDownload ? ['far', 'trash'] : ['fas', 'cloud-download']} key="icon" />
            let buttonClass = this.state.isSelectedForDownload ? 'off' : 'on';
            let textContent = React.createElement('span', { key: "textContent" }, this.state.isSelectedForDownload ? 'Fjern nedlasting' : 'Last ned')

            let childElements = [icon, textContent];
            return React.createElement('span', { onClick: action, className: buttonClass }, childElements);

        }
        else if (this.showDownloadLink()) {
            let distributionUrl = this.props.searchResult.DistributionUrl
            let icon = <FontAwesomeIcon icon={['far', 'download']} key="icon" />
            let buttonClass = this.state.isSelectedForDownload ? 'off' : 'on';
            let textContent = React.createElement('span', { key: "textContent" }, 'Til nedlasting')

            let childElements = [icon, textContent];
            return React.createElement('a', { href: distributionUrl, className: buttonClass }, childElements);
        }
        else {
            let content = '';
            let buttonClass = 'btn btn-sm disabled';
            return React.createElement('span', { className: buttonClass }, content);
        }
    }
}

DownloadButton.propTypes = {
    searchResult: PropTypes.object.isRequired
}

const mapStateToProps = state => ({

});

const mapDispatchToProps = {
    removeItemSelectedForDownload,
    addItemSelectedForDownload
};

export default connect(mapStateToProps, mapDispatchToProps)(DownloadButton);