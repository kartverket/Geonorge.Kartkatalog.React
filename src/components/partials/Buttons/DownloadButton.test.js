import React from 'react';
import {shallow} from "enzyme";
import {DownloadButton} from './DownloadButton';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import style from './Buttons.scss'

function setupMetadataDownloadButton() {
    const metadata = {
        DistributionProtocol: "GEONORGE:DOWNLOAD"
    };
    const itemsToDownload = [];
    const props = {
        metadata: metadata,
        itemsToDownload: itemsToDownload,
        addItemSelectedForDownload: jest.fn(),
        removeItemSelectedForDownload: jest.fn(),
        getResource: jest.fn()
    };
    const wrapper = shallow(<DownloadButton {...props} />);

    return {
        props,
        wrapper
    }
}

function setupMetadataDownloadLink() {
    const metadata = {
        DistributionUrl: "test.no",
        DistributionProtocol: "WWW:DOWNLOAD-1.0-http--download",
        Type: "dataset"
    };
    const itemsToDownload = [];
    const props = {
        metadata: metadata,
        itemsToDownload: itemsToDownload,
        addItemSelectedForDownload: jest.fn(),
        removeItemSelectedForDownload: jest.fn(),
        getResource: jest.fn()
    };
    const wrapper = shallow(<DownloadButton {...props} />);

    return {
        props,
        wrapper
    }
}

function setupMetadataListButtonFalse() {
    const metadata = {
    };
    const itemsToDownload = [];
    const props = {
        metadata: metadata,
        itemsToDownload: itemsToDownload,
        addItemSelectedForDownload: jest.fn(),
        removeItemSelectedForDownload: jest.fn(),
        getResource: jest.fn(),
        listButton: false
    };
    const wrapper = shallow(<DownloadButton {...props} />);

    return {
        props,
        wrapper
    }
}

function setupMetadataListButtonFalseAndCanShowDownloadService() {
    const metadata = {
        CanShowDownloadService: true
    };
    const itemsToDownload = [];
    const props = {
        metadata: metadata,
        itemsToDownload: itemsToDownload,
        addItemSelectedForDownload: jest.fn(),
        removeItemSelectedForDownload: jest.fn(),
        getResource: jest.fn(),
        listButton: false
    };
    const wrapper = shallow(<DownloadButton {...props} />);

    return {
        props,
        wrapper
    }
}

function setupMetadataListButtonFalseAndCanShowDownloadUrl() {
    const metadata = {
        CanShowDownloadUrl: true,
        DistributionUrl: 'test.no'
    };
    const itemsToDownload = [];
    const props = {
        metadata: metadata,
        itemsToDownload: itemsToDownload,
        addItemSelectedForDownload: jest.fn(),
        removeItemSelectedForDownload: jest.fn(),
        getResource: jest.fn(),
        listButton: false
    };
    const wrapper = shallow(<DownloadButton {...props} />);

    return {
        props,
        wrapper
    }
}


describe('DownloadButton', () => {


    // List button
    it('should render self', () => {
        const {wrapper} = setupMetadataDownloadButton();
        expect(wrapper).toMatchSnapshot();
    });

    it('Show download button when item is selected for download', () => {
        let {wrapper} = setupMetadataDownloadButton();
        wrapper = wrapper.setState({isAdded: true});

        expect(wrapper.prop("className")).toContain("off");
        expect(wrapper.prop("onClick").toString()).toContain("removeFromDownloadList(button)");
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('trash');
        expect(wrapper.find("span").first().html()).toContain('Fjern nedlasting');
    });

    it('Show download button when item is not selected for download', () => {
        let {wrapper} = setupMetadataDownloadButton();
        wrapper = wrapper.setState({isAdded: false});

        expect(wrapper.prop("className")).toContain("on");
        expect(wrapper.prop("onClick").toString()).toContain("addToDownloadList(button)");
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('cloud-download');
        expect(wrapper.find("span").first().html()).toContain('Last ned');
    });

    it('Show download link', () => {
        let {wrapper} = setupMetadataDownloadLink();

        expect(wrapper.prop("className")).toContain("on");
        expect(wrapper.prop("href")).toBe('test.no');
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('external-link-square');
        expect(wrapper.find("span").first().html()).toContain('Til nedlasting');
    });

    
    // Button

    it('List button false', () => {
        let {wrapper} = setupMetadataListButtonFalse();

        expect(wrapper.prop("className")).toContain(style.btn + ' disabled')
        expect(wrapper.prop("href")).toBeUndefined()
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('cloud-download');
        expect(wrapper.find("span").first().html()).toContain('Last ned');
    });

    it('List button false - CanShowDownloadService true', () => {
        let {wrapper} = setupMetadataListButtonFalseAndCanShowDownloadService();
        wrapper = wrapper.setState({isAdded: false});

        expect(wrapper.prop("className")).toContain(style.btn + ' download')
        expect(wrapper.prop("onClick").toString()).toContain("addToDownloadList(button)");
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('cloud-download');
        expect(wrapper.find("span").first().html()).toContain('Last ned');
    });

    it('List button false - CanShowDownloadUrl true', () => {
        let {wrapper} = setupMetadataListButtonFalseAndCanShowDownloadUrl();

        expect(wrapper.prop("className")).toContain(style.btn)
        expect(wrapper.prop("href")).toBe('test.no');
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('external-link-square');
        expect(wrapper.find("span").first().html()).toContain('Til nedlasting');
    });
});
