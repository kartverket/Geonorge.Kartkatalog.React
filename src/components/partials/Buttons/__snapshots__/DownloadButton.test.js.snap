// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadButton should render self 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DownloadButton
    addItemSelectedForDownload={[MockFunction]}
    getResource={
      [MockFunction] {
        "calls": Array [
          Array [
            "Download",
            "Last ned",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    itemsToDownload={Array []}
    listButton={true}
    metadata={
      Object {
        "DistributionProtocol": "GEONORGE:DOWNLOAD",
      }
    }
    removeItemSelectedForDownload={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Array [
              "fas",
              "cloud-download",
            ]
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />,
        <span />,
      ],
      "className": "on",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "icon",
        "nodeType": "function",
        "props": Object {
          "border": false,
          "className": "",
          "fixedWidth": false,
          "flip": null,
          "icon": Array [
            "fas",
            "cloud-download",
          ],
          "inverse": false,
          "listItem": false,
          "mask": null,
          "pull": null,
          "pulse": false,
          "rotation": null,
          "size": null,
          "spin": false,
          "symbol": false,
          "title": "",
          "transform": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "textContent",
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "span",
      },
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Array [
                "fas",
                "cloud-download",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />,
          <span />,
        ],
        "className": "on",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "icon",
          "nodeType": "function",
          "props": Object {
            "border": false,
            "className": "",
            "fixedWidth": false,
            "flip": null,
            "icon": Array [
              "fas",
              "cloud-download",
            ],
            "inverse": false,
            "listItem": false,
            "mask": null,
            "pull": null,
            "pulse": false,
            "rotation": null,
            "size": null,
            "spin": false,
            "symbol": false,
            "title": "",
            "transform": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "textContent",
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "span",
        },
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
